### PR TITLE
fix(semantic): implicit return `UpdateExpression` in `ArrowFunctionExpression` does not as read reference

### LINKED ISSUE #5156: bug(semantic): implicit return in arrow function does not count as a read
Given a variable `i`:

Expected behavior
image

Actual behavior
image

This causes `no-unused-vars` to incorrectly report `i` as unused.

### PR DESCRIPTION
**Pull request by Dunqing**
close: #5158
close: #5156

### CODE REVIEW

**Comment by Dunqing** [comment]
* **#5161** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/5161?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Dunqing and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: crates/oxc_semantic/src/builder.rs [hunk 1]
```diff
@@ -1765,6 +1765,18 @@ impl<'a> SemanticBuilder<'a> {
                 self.function_stack.push(self.current_node_id);
                 self.make_all_namespaces_valuelike();
             }
+            AstKind::FunctionBody(_) => {
```
Comment (DonIsaac) [inline crates/oxc_semantic/src/builder.rs:1768]: Can you add test cases for `let i = 1; const foo = () => () => { i++ }`?

Comment (Dunqing) [reply crates/oxc_semantic/src/builder.rs:1768]: The current implementation broke this test 🥲. I add and fix it in https://github.com/oxc-project/oxc/pull/5161/commits/66a1b9d4680f8c35a9d4d4e8d05d661a8dd6c74f

File: crates/oxc_semantic/src/builder.rs [hunk 2]
```diff
@@ -1765,6 +1765,23 @@ impl<'a> SemanticBuilder<'a> {
...
+                    if arrow.expression {
+                        // (x) => x++
+                        //        ^^^ when the arrow function is an expression,
+                        //            the `x++` is a implicit return,
+                        //            so we need to count `x` as a read reference
+                        self.current_reference_flags |= ReferenceFlags::Read;
+                    } else {
+                        // (x) => () => { x++ }
+                        //                ^^^ avoid break nested arrow function is not an expression
+                        self.current_reference_flags -= ReferenceFlags::Read;
```
Comment (DonIsaac) [inline crates/oxc_semantic/src/builder.rs:1781]: I've got a feeling this could add problems down-stream... this is why I added an `is_in_return` flag

**Comment by Boshen** [comment]
This blocks us from making a release.

File: crates/oxc_semantic/src/builder.rs (hunk 2 above)
Comment (Dunqing) [reply crates/oxc_semantic/src/builder.rs]: I've tried another way, please review it again

**Review by DonIsaac — APPROVED** [review]

**Pull request merged**